python/gtsam/examples/CameraResectioning.py

```diff
@@ -0,0 +1,85 @@
+# pylint: disable=consider-using-from-import,invalid-name,no-name-in-module,no-member,missing-function-docstring
+"""
+This is a 1:1 transcription of CameraResectioning.cpp.
+"""
+import numpy as np
+from gtsam import Cal3_S2, CustomFactor, LevenbergMarquardtOptimizer, KeyVector
+from gtsam import NonlinearFactor, NonlinearFactorGraph
+from gtsam import PinholeCameraCal3_S2, Point2, Point3, Pose3, Rot3, Values
+from gtsam.noiseModel import Base as SharedNoiseModel, Diagonal
+from gtsam.symbol_shorthand import X
+
+
+def resectioning_factor(
+    model: SharedNoiseModel,
+    key: int,
+    calib: Cal3_S2,
+    p: Point2,
+    P: Point3,
+) -> NonlinearFactor:
+
+    def error_func(this: CustomFactor, v: Values, H: list[np.ndarray]) -> np.ndarray:
+        pose = v.atPose3(this.keys()[0])
+        camera = PinholeCameraCal3_S2(pose, calib)
+        if H is None:
+            return camera.project(P) - p
+        Dpose = np.zeros((2, 6), order="F")
```

Comment:
This is correct, as these are passed as reference to C++. Please check `pybind11`'s documentation on this.

Reply:
Thanks!  This is what I'd expect given the way pybind seems to work.

The mystery is: what the CustomFactor is doing to allow either "C" or "F" storage for H[0]?

Oh also the cheirality issue I'm seeing doesn't have anything to do with this python code at all, it happens with the CameraResectioning.cpp too.  It seems to be that the optimizer is simply picking negative z's on its way to the solution.  Picking an initial Z of 5, for example, will trigger the exception as the optimizer happily slews around into negative territory.  Turning GTSAM_THROW_CHEIRALITY_EXCEPTION off fixes it; is there a downside to running that way, assuming there are enough points to avoid finding *solutions* on the wrong side of the camera?

Reply:
It will copy if you give something with wrong storage order, simple and effective :)

The 2nd question is probably not related to Python though. Maybe just catch it in Python would work?

Reply:
> This is correct, as these are passed as reference to C++. Please check `pybind11`'s documentation on this.

Uhm, Fan, my ask to you is: you (as the author of this amazing capability) read the pybind11 docs and make sure the readme file has everything a casual user of GTSAM (with no knowledge of pybind necessary) can use it by following instructions :-). I think the readme currently does not mention the storage order, so this is what I'm asking you to fix (and check the examples) if you're willing.

Reply:
> The 2nd question is probably not related to Python though. Maybe just catch it in Python would work?

yeah, that's right, it's not a python issue, just something i noticed while working with the camera resectioning example.  you can't just catch it, because you need the optimizer to traverse the 'bad' territory on the way to the correct answer.  in this very simple case, the optimizer walks the "z" axis from the initial value of 5 to the correct value of 2 along this path: [5, -2.5, -2.49, -2.48, -2.35, -1.27, 2.46, 1.9, 1.99, 2.0].

Reply:
Try using LM with a high initial value for lambda

Reply:
thanks!  that works, i should have thought of it.  would you recommend, in general, the "high lambda" solution over the "don't throw chirality exception" solution?

Reply:
Oh I mean since you are calling `PinholeCameraCal3_S2` in Python, it is possible that you can catch the exception here in your error function and do special processing.

> readme currently does not mention the storage order

I see. The gist is that whatever thing that got passed to C++ as a **reference to write into** has to be in the "F" format. This is because the buffer will be directly written by C++ code which assumes Eigen (column) layout. If a matrix in C++ got passed to Python for writing, then the ordering matters less, as Pybind11 will do automatic copy when you assign to the matrix with a numpy matrix with row-major layout. There is a small performance penalty though. It's best that "F" matrix is used at all times.